packages/flutter_test/lib/src/finders.dart

```diff
@@ -520,121 +907,225 @@ abstract class Finder {
-    if (result.isNotEmpty) {
-      _cachedResult = result;
+    assert(_precacheResults == null);
+    if (tryFind()) {
       return true;
     }
-    _cachedResult = null;
+    _precacheResults = null;
     return false;
   }
 
-  /// Returns a variant of this finder that only matches the first element
-  /// matched by this finder.
-  Finder get first => _FirstFinder(this);
+  @override
+  Iterable<Element> findInCandidates(Iterable<Element> candidates) {
+    return apply(candidates);
+  }
 
-  /// Returns a variant of this finder that only matches the last element
-  /// matched by this finder.
-  Finder get last => _LastFinder(this);
+  @override
+  String toString({bool describeSelf = false}) {
+    final String result = super.toString(describeSelf: describeSelf);
+    _found = null;
+    return result;
+  }
+}
 
-  /// Returns a variant of this finder that only matches the element at the
-  /// given index matched by this finder.
-  Finder at(int index) => _IndexFinder(this, index);
+/// A base class for creating finders that search the [Element] tree for
+/// [Widget]s.
+abstract class WidgetFinder extends FinderBase<Element> with _LegacyFinderMixin {
+  /// Creates a new [WidgetFinder] with the given [skipOffstage] value.
+  WidgetFinder({this.skipOffstage = true});
+
+  /// Whether this finder skips nodes that are offstage.
+  ///
+  /// If this is true, then the elements are walked using
+  /// [Element.debugVisitOnstageChildren]. This skips offstage children of
+  /// [Offstage] widgets, as well as children of inactive [Route]s.
+  @override
+  final bool skipOffstage;
+
+  @override
+  WidgetFinder get first => _FirstWidgetFinder(this);
+
+  @override
+  WidgetFinder get last => _LastWidgetFinder(this);
+
+  @override
+  WidgetFinder at(int index) => _IndexWidgetFinder(this, index);
+
+  @override
+  Iterable<Element> get allCandidates {
+    return collectAllElementsFrom(
+      WidgetsBinding.instance.rootElement!,
+      skipOffstage: skipOffstage,
+    );
+  }
+
+  @override
+  String describeCandidate(Plurality plurality) {
+    switch (plurality) {
+      case Plurality.zero:
+      case Plurality.many:
+        return 'widgets with $description';
+      case Plurality.one:
+        return 'widget with $description';
+    }
+  }
 
   /// Returns a variant of this finder that only matches elements reachable by
   /// a hit test.
   ///
   /// The [at] parameter specifies the location relative to the size of the
   /// target element where the hit test is performed.
-  Finder hitTestable({ Alignment at = Alignment.center }) => _HitTestableFinder(this, at);
+  WidgetFinder hitTestable({ Alignment at = Alignment.center }) => _HitTestableWidgetFinder(this, at);
+}
+
+/// A base class for creating finders that search the semantics tree.
+abstract class SemanticsFinder extends FinderBase<SemanticsNode> {
+  /// Creates a new [SemanticsFinder] that will search starting at the given
+  /// [root].
+  SemanticsFinder(this.root);
```

Comment:
I can't define a SemanticsFinder that gives me - say - all buttons labeled "Foo" across all semantics trees? I kind of expected that if you don't specify a view/root, the finder would just search across views.

Reply:
Could you expand on why you would expect that behavior? The trees are all separate for a reason, so it would make sense to me that they would be searched separately. I'd worry about introducing extra edge cases by trying to search all trees if we did it that was as well.

Reply:
The fact that the trees are separate is kinda an implementation detail. If I want to find a tappable semanticsnode with label "foo" I shouldn't need to think about what tree or view it is in.

That's how widget finders work as well: I can just say "give me all buttons with text `foo`" and I will get them across all views. I don't have to think about what view they are in (or whether all widgets are in the same tree or not)

Reply:
I'd argue that it's _not_ just an implementation detail. When writing or interacting with the app, it'll be obvious that the semantics trees are separate, so I would expect the testing tools to behave the same personally.

The semantics finders API was designed assuming multiple trees, and that the specific tree would have to be provided (via a FlutterView using tester.viewOf) from the start if you weren't using the implicit view provided by the testing environment. The semantics side would have to be pretty heavily reworked to support searching multiple trees.

Aside from my own expectations and the additional rework, I'd be concerned about inconsistency and the risk of false positives.

Both sets of finders are pretty consistent right now, because they navigate a single tree.  I'd worry that navigating multiple trees would potentially introduce inconsistencies if the order of the roots were to change. I could also see the ancestor/descendant finders _feeling_ inconsistent because you'd end up with ancestors/descendants across multiple trees, which seems counterintuitive with the idea of those finders.

As far as the false positives, having the same/similar elements on multiple windows isn't that unlikely, and semantics are more limited than visual UIs, so that chance of that is even worse. We'd have to have _a_ way to limit the search to a specific tree. As it is right now, it's built into everything. If the default behavior changes to search all trees, then we'd need to add something for finding the semantics root of a specific view, and we would then have to expect users to use descendant finders with that new method to limit their search.

Personally, I think the explicit separate trees is the correct approach. If this is something that you feel _must_ be changed, let me know and I'll get it to work, I just don't think it's the better approach or experience.

Reply:
Per in-call discussions, for now the plan is to keep the single tree searching behavior. Primarily because the order of the SemanticsNode roots is not guaranteed and could be seen as "unpredictable" from a developer point of view, even if it is technically deterministic.